applications/crossbar/src/modules/cb_tasks.erl

```diff
@@ -0,0 +1,433 @@
+%% @end
+%%--------------------------------------------------------------------
+-spec content_types_accepted(cb_context:context(), path_token()) -> cb_context:context().
+content_types_accepted(Context, ?HELP) ->
+    Context;
+content_types_accepted(Context, _TaskId) ->
+    cta(Context, cb_context:req_verb(Context)).
+
+-spec cta(cb_context:context(), http_method()) -> cb_context:context().
+cta(Context, ?HTTP_PUT) ->
+    CTA = [{'from_binary', ?CSV_CONTENT_TYPES ++ ?JSON_CONTENT_TYPES}],
+    cb_context:add_content_types_accepted(Context, CTA);
+cta(Context, _) ->
+    Context.
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc
+%% What content-types will the module be using to respond (matched against
+%% client's Accept header)
+%% Of the form {atom(), [{Type, SubType}]} :: {to_json, [{<<"application">>, <<"json">>}]}
+%% @end
+%%--------------------------------------------------------------------
+-spec content_types_provided(cb_context:context(), path_token()) -> cb_context:context().
+content_types_provided(Context, ?HELP) ->
+    Context;
+content_types_provided(Context, _TaskId) ->
+    ctp(Context, cb_context:req_verb(Context)).
+
+-spec ctp(cb_context:context(), http_method()) -> cb_context:context().
+-spec ctp(cb_context:context()) -> cb_context:context().
+ctp(Context, ?HTTP_GET) ->
+    ctp(Context);
+ctp(Context, _) ->
+    Context.
+ctp(Context) ->
+    CTP = [{'to_json', ?JSON_CONTENT_TYPES}
+          ,{'to_binary', ?CSV_CONTENT_TYPES}
+          ],
+    cb_context:add_content_types_provided(Context, CTP).
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc
+%% Check the request (request body, query string params, path tokens, etc)
+%% and load necessary information.
+%% /tasks mights load a list of task objects
+%% /tasks/123 might load the task object 123
+%% Generally, use crossbar_doc to manipulate the cb_context{} record
+%% @end
+%%--------------------------------------------------------------------
+-spec validate(cb_context:context()) -> cb_context:context().
+-spec validate(cb_context:context(), path_token()) -> cb_context:context().
+validate(Context) ->
+    validate_tasks(Context, cb_context:req_verb(Context)).
+validate(Context, PathToken) ->
+    validate_task(Context, PathToken, cb_context:req_verb(Context)).
+
+-spec validate_tasks(cb_context:context(), http_method()) -> cb_context:context().
+validate_tasks(Context, ?HTTP_GET) ->
+    summary(Context);
+validate_tasks(Context, ?HTTP_PUT) ->
+    QS = cb_context:query_string(Context),
+    case {kz_json:get_ne_binary_value(?QS_CATEGORY, QS)
+         ,kz_json:get_ne_binary_value(?QS_ACTION, QS)
+         }
+    of
+        {?NE_BINARY, ?NE_BINARY} ->
+            validate_new_attachment(Context, is_content_type_csv(Context));
+        {_, _} -> cb_context:add_system_error('invalid request', Context)
```

Comment:
I wonder if there's a way to provide better feedback here. Why not create schema errors? And why restrict category and action to the query string? `cb_context:req_value/2` will check request body and query string.